Redis Programmability Documentation (#1725)
redis/redis-doc: topics/lua-api.md

Scripts **should never** access keys with programmatically-generated names or based on the contents of data structures stored in the database.

The _KEYS_ global variable is available only for [ephemeral scripts](/topics/eval-intro).
It is pre-populated with all key name input arguments.

### The _ARGV_ global variable

* Since version: 2.6.0
* Available in scripts: yes
* Available in functions: no

The _ARGV_ global variable is available only in [ephemeral scripts](/topics/eval-intro).
It is pre-populated with all regular input arguments.

## _redis_ object

* Since version: 2.6.0
* Available in scripts: yes
* Available in functions: yes

The Redis Lua execution context always provides a singleton instance of an object named _redis_.
The _redis_ instance enables the script to interact with the Redis server that's running it.
Following is the API provided by the _redis_ object instance.

### `redis.call(command [,arg...])`
<a href="redis.call()"></a>

* Since version: 2.6.0
* Available in scripts: yes
* Available in functions: yes

The `redis.call()` function calls a given Redis command and returns its reply.

> Its inputs are the command and arguments, and once called, it executes the command in Redis and returns the reply.

Reviewer: i think this section (speaking about `redis.call` is missing some link to where we talk about OOM errors

redis.call('SET', KEYS[1], ARGV[1])
redis.set_repl(redis.REPL_NONE)
redis.call('SET', KEYS[2], ARGV[2])
redis.set_repl(redis.REPL_ALL)
redis.call('SET', KEYS[3], ARGV[3])
```

If you'll run this script by calling `EVAL "..." 3 A B C 1 2 3`, the result will be that only the keys _A_ and _C_ are be created on the replicas and AOF.

### `redis.replicate_commands()`

* Since version: 3.2.0
* Until version: 7.0.0
* Available in scripts: yes
* Available in functions: no

This function switches the script's replication mode from verbatim replication to effects replication.
You can use it to override the default verbatim script replication mode used by Redis until version 7.0.

**Note:**
as of Redis v7.0, verbatim script replication is no longer supported.
The default, and only script replication mode supported, is script effects' replication.
For more information, please refer to [`Replicating commands instead of scripts`](eval-intro#replicating-commands-instead-of-scripts)

### `redis.breakpoint()`

* Since version: 3.2.0
* Available in scripts: yes
* Available in functions: no

This function triggers a breakpoint when using the Redis Lua debugger](/topics/ldb).

> ### `redis.debug(x)`

Reviewer: redis.VERSION and redis.VERSION_NUM are missing, right?

This function prints its argument in the [Redis Lua debugger](/topics/ldb) console.

### <a name="redis.register_function()"></a> `redis.register_function`

* Since version: 7.0.0
* Available in scripts: no
* Available in functions: yes

This function is only available from the context of the `FUNCTION LOAD` command.
When called, it registers a function to the loaded library.
The function can be called either with positional argument or with named arguments.

#### <a name="redis.register_function_pos_args"></a> positional arguments: `redis.register_functio(name, callback)`

The first argument to `redis.register_functio` is a Lua string representing the function name.
The second argument to `redis.register_functio` is a Lua function.

Error raise on the following cases:

* Wrong number of arguments
* Wrong argument type

> * Another function with the same name already registered on the current library

Reviewer: i don't think we need to go to that level of detail (considering the user gets a sensible error when it happens)

The function can be called either with positional or named arguments.

#### <a name="redis.register_function_pos_args"></a> positional arguments: `redis.register_functio(name, callback)`

The first argument to `redis.register_function` is a Lua string representing the function name.
The second argument to `redis.register_function` is a Lua function.

Usage example:

```
redis> FUNCTION LOAD Lua mylib "redis.register_function('noop', function() end)"
```

#### <a name="redis.register_function_named_args"></a> Named arguments:  `redis.register_functio{function_name=name, callback=callback, flags={flag1, flag2, ..}. description=description}`

The named arguments variant accepts the following arguments:

* _function\_name_: the function's name.
* _callback_: the function's callback.
* _flags_: an array of strings, each a function flag (optional).
* _description_: function's description (optional).

Both _function\_name_ and _callback_ are mandatory.

Usage example:

```
redis> FUNCTION LOAD Lua mylib "redis.register_function(function_name='noop', callback=function() end, flags={ 'no-writes' }, description='Does nothing')"
```

> #### <a name="function_flags"></a> Function flags

Reviewer: flags are now serving EVAL scripts too, so maybe we should call the section "script flags"
and have a short intro that states (links) how you apply them to either Eval or Function.

we can maybe leave the details about EVAL to be added later.
but let's avoid calling them "function flags" in all the text below, and the section name.

**Important:**
use function flags with care, which may negatively impact if misused.

When you register a function, the server does not know how it accesses the database.
By default, Redis assumes that all functions read and write data.
This results in the following behavior:

1. Functions can read and write data.
1. Functions can run in cluster mode.
1. Execution against a stale replica is denied to avoid inconsistent reads.

> 1. Execution under low memory is denied to avoid exceeding the configured threshold.

Reviewer 2: Add that it's denied in a read-only replica?

You can use the following flags and instruct the server to treat the function's execution differently:

* `no-writes`: this flag indicates that the function only reads data but never writes.

    By default, Redis will deny the execution of functions against read-only replicas, as they may attempt to perform writes.
    Similarly, the server will not allow calling functions with `FCALL_RO`.
    Lastly, when data persistence is at risk due to a disk error, execution is blocked as well.
    
    Using this flag allows executing the function:

>     1. With `FCALL_RO` against masters and read-only replicas.

Reviewer 2: ```suggestion

    1. With `FCALL_RO` against masters and read-only replicas.
```

    2. Even if there's a disk error.

    However, note that the server will return an error if the function attempts to call a write command.

* `allow-oom`: use this flag to allow a function to execute when the server is out of memory (OOM).
  
    Unless used, Redis will deny the execution of functions when in an OOM state, regardless of the `no-write` flag and method of calling.
    Furthermore, when you use this flag, the function can call any Redis command, including commands that aren't usually allowed in this state.

* `allow-stale`: a flag that enables running the function against a stale replica.

    By default, Redis prevents data consistency problems from using old data by having stale replicas return a runtime error.

>     In cases where the consistency is a lesser concern, this flag allows stale Redis replicas to run the function.

Reviewer: is stale replica a "consistency" issue? i'm not sure.

* _flags_ - Lua array of strings representing function flags
* _description_ - Lua string representing the function description

_function\_name_ and _callback_ are mandatory, the other named arguments are option.

Usage Example:

```
redis> FUNCTION LOAD Lua mylib "redis.register_function{function_name='noop', callback=function() end, function, flags={'no-writes'}, description='some desc')"
```

#### <a name="script_flags"></a> Script Flags

> The following is a list of supported script flags:

Reviewer: i think we should mention the default.
i.e. on functions, all these flags are off by default.
on eval... when shebang is used, it's the same, but if it's not used, it's a different story (maybe refer to another section)
@yoav-steinberg FYI

> * `no-writes` - indicating the function perform no writes which means that it is OK to run it on:

Reviewer: ```suggestion
* `no-writes` - indicating the script perform no writes which means that it is OK to run it on:
```

   * read-only replica

>    * Using FCALL_RO

Reviewer: ```suggestion
   * Using FCALL_RO / EVAL_RO
```

>    * If disc error detected

Reviewer: ```suggestion
   * If disk error detected (Redis is unable to persist so it rejects writes)
```

>    It will not be possible to run a function in those situations unless the function turns on the `no-writes` flag, in this case Redis will inforce the function is doing not writes and will raise an error if it does.

Reviewer: ```suggestion
   It will not be possible to run a function in those situations unless the script turns on the `no-writes` flag, in this case Redis will enforce the script is doing not writes and will raise an error if it does.
```

> * `allow-oom` - indicate that its OK to run the function even if Redis is in OOM (out of memory) state, if the function will not turn on this flag it will not be possible to run it if OOM reached (even if the function declares `no-writes` and even if `fcall_ro` is used). If this flag is set, any command will be allowed on OOM (even ones that are normally not allowed in that state).

Reviewer: ```suggestion
* `allow-oom` - indicate that it's OK to run the script even if Redis is in OOM (out of memory) state, if the script will not turn on this flag it will not be possible to run it if OOM reached (even if the script declares `no-writes` and even if `fcall_ro` / `eval_ro` is used). If this flag is set, any command will be allowed on OOM (even ones that are normally not allowed in that state).
```

> * `allow-state` - indicate that its OK to run the function on stale replica, in this case calls to Redis commands that are not allowed in this state will error when the script is executed on a stale replica.

Reviewer: ```suggestion
* `allow-state` - indicate that it's OK to run the script on stale replica, in this case calls to Redis commands that are not allowed in that state will error when the script is executed on a stale replica.
```

> * `no-cluster` - indicate to disallow running the function on Redis cluster. 

Reviewer: ```suggestion
* `no-cluster` - indicate to disallow running the script on Redis cluster. 
```
